ember-resources/src/util/function.ts

```diff
@@ -54,118 +45,79 @@ type UseFunctionArgs<Return> = Vanilla<Return> | WithInitialValue<Return>;
- * @param {Function} theFunction the function to run with the return value available on .value
- */
-export function trackedFunction<Return>(...passed: WithInitialValue<Return>): State<Return>;
-
-export function trackedFunction<Return>(...passedArgs: UseFunctionArgs<Return>) {
-  let [context] = passedArgs;
-  let initialValue: Return | undefined;
-  let fn: ResourceFn<Return>;
-
-  assert(
-    `Expected second argument to useFunction to either be an initialValue or the function to run`,
-    passedArgs[1] !== undefined
-  );
-
-  if (hasNoInitialValue(passedArgs)) {
-    fn = passedArgs[1];
-  } else {
-    initialValue = passedArgs[1];
-    fn = passedArgs[2];
-  }
-
-  return resource<State<Return>>(context, (hooks) => {
-    let state = new State(fn, hooks, initialValue);
+export function trackedFunction<Return>(context: object, fn: () => Return) {
+  const state = new State(fn);
 
+  let destroyable = resource<State<Return>>(context, () => {
     state.retry();
 
     return state;
   });
+
+  associateDestroyableChild(destroyable, state);
+
+  return destroyable;
 }
 
 /**
  * State container that represents the asynchrony of a `trackedFunction`
  */
 export class State<Value> {
-  @tracked isResolved = false;
-  @tracked resolvedValue?: Value;
-  @tracked error?: unknown;
+  @tracked data: TrackedAsyncData<Value> | null = null;
+  @tracked declare promise: Value;
 
-  #fn: ResourceFn<Value>;
-  #hooks: Hooks;
-  #initialValue: Value | undefined;
+  #fn: () => Value;
 
-  constructor(fn: ResourceFn<Value>, hooks: Hooks, initialValue?: Value) {
+  constructor(fn: () => Value) {
     this.#fn = fn;
-    this.#hooks = hooks;
-    this.#initialValue = initialValue;
   }
 
-  get value() {
-    return this.resolvedValue || this.#initialValue || null;
+  get state(): 'UNSTARTED' | 'PENDING' | 'RESOLVED' | 'REJECTED' {
+    return this.data?.state ?? 'UNSTARTED';
   }
 
   get isPending() {
-    return !this.isResolved;
+    if (!this.data) return true;
+
+    return this.data.isPending ?? false;
+  }
+
+  get isResolved() {
+    return this.data?.isResolved ?? false;
   }
 
-  get isLoading() {
-    return this.isPending;
+  get isRejected() {
+    return this.data?.isRejected ?? false;
   }
 
-  get isError() {
-    return Boolean(this.error);
+  /**
+   * TrackedAsyncData does not allow the accessing of data before
+   * .state === 'RESOLVED'  (isResolved).
+   *
+   * From a correctness standpoint, this is perfectly reasonable,
+   * as it forces folks to handle thet states involved with async functions.
+   *
+   * The original version of `trackedFunction` did not use TrackedAsyncData,
+   * and did not have these strictnesses upon property access, leaving folks
+   * to be as correct or as fast/prototype-y as they wished.
+   *
+   * For now, `trackedFunction` will retain that flexibility.
+   */
+  get value(): Awaited<Value> | null {
+    if (this.data?.isResolved) {
```

Comment:
Why is this necessary?

You say "TrackedAsyncData does not allow the accessing of data before `.state === 'RESOLVED'`".

But in `ember-async-data` types, the `data` property  is typed to be `null` for `PENDING` and `ERROR` states: https://github.com/tracked-tools/ember-async-data/blob/dbd10ab23395379494386e3892b0a04343937541/ember-async-data/src/tracked-async-data.ts#L271

Accessing `data.value` is not disallowed in `ember-async-data`, and I in my PR I [have](https://github.com/NullVoxPopuli/ember-resources/pull/683/files?w=1#diff-bb529aecf02ddc83e5aa05bf632d7d6c140df5e368b9ccd1c61a16a5ac3f6453R105) tests that access `data.value` without issue.

Reply:
I need to update the comment, the real issue is that `this.data` might be undefined, due to when the TrackedAsyncData instance is created

Reply:
oh no I was running in to this: https://github.com/tracked-tools/ember-async-data/blob/v0.6.0/addon/tracked-async-data.ts#L150

Reply:
Oh, that's annoying! I really don't think that preventing access to `value` is necessary.

Thx for working around it!